core/src/main/java/hudson/PluginManager.java

```diff
@@ -650,7 +650,17 @@ protected static void addDependencies(URL hpiResUrl, String fromPath, Set<URL> d
                     continue;
                 }
 
-                String artifactId = dependencyToken.split(":")[0];
+                String[] artifactIdVersionPair = dependencyToken.split(":");
+                String artifactId = artifactIdVersionPair[0];
+                VersionNumber dependencyVersion = new VersionNumber(artifactIdVersionPair[1]);
+
+                PluginManager manager = Jenkins.getActiveInstance().getPluginManager();
+                VersionNumber installedVersion = manager.getPluginVersion(manager.rootDir, artifactId);
+                if (installedVersion != null && installedVersion.isNewerThan(dependencyVersion)) {
```

Comment:
This only checks that we don't downgrade plugins that are already installed. I think this code is still broken in the case that 2 detached plugins depend on different versions of the same plugin.

Reply:
In that case, I'm not sure which version of the dependency gets installed.

Reply:
> I think this code is still broken in the case that 2 detached plugins depend on different versions of the same plugin.

Seems so. Looking into the code to see how we could fix that.

Reply:
The only thing I can think of would be to introduce a class with the plugin name, version, and url, with equality based only on the name, and use it instead of `Set<URL>` so that we can check versions as we add to the set.

Reply:
Yes, it's what I think as well. We had pretty similar issues in https://github.com/jenkinsci/jenkins/pull/2903 with @lvotypko for the `UpdateSite`. Maybe we could somehow align approaches, because the dependency management becomes a real problem

Reply:
Note that we should make all changes very carefully, because there are other `PluginManager` implementations depending on the current protected APIs